### PR TITLE
[BOLT][NFC] Refactor function matching

### PR DESCRIPTION
**Pull request by shawbyoung**
Moved function matching techniques into separate helper functions for
ease of understanding and to make space for additional function 
matching techniques to be added (e.g. call graph function matching).

### CODE REVIEW

File: bolt/include/bolt/Profile/YAMLProfileReader.h [hunk 1]
```diff
@@ -73,13 +73,26 @@ class YAMLProfileReader : public ProfileReaderBase {
   bool parseFunctionProfile(BinaryFunction &Function,
                             const yaml::bolt::BinaryFunctionProfile &YamlBF);
 
+  /// Returns block cnt equality if IgnoreHash is true, otherwise, hash equality
+  bool profileMatches(const yaml::bolt::BinaryFunctionProfile &Profile,
```
Comment (dcci) [inline bolt/include/bolt/Profile/YAMLProfileReader.h:77]: I think this comment talks about the implementation more than the definition. Can you rephrase it?

File: bolt/include/bolt/Profile/YAMLProfileReader.h [hunk 2]
```diff
@@ -73,13 +73,26 @@ class YAMLProfileReader : public ProfileReaderBase {
...
+
   /// Infer function profile from stale data (collected on older binaries).
   bool inferStaleProfile(BinaryFunction &Function,
                          const yaml::bolt::BinaryFunctionProfile &YamlBF);
 
   /// Initialize maps for profile matching.
   void buildNameMaps(BinaryContext &BC);
 
+  /// Matches functions using exact name.
+  size_t matchWithExactName();
```
Comment (dcci) [inline bolt/include/bolt/Profile/YAMLProfileReader.h:88]: why you need these 3 different functions?

**Review by dcci — COMMENTED** [review]
I have a couple of general comments about this.
Can you also please add a description explaining what this patch does?

**Comment by dcci** [comment]
> I have a couple of general comments about this. Can you also please add a description explaining what this patch does?

i.e. why we're refactoring these functions.

File: bolt/include/bolt/Profile/YAMLProfileReader.h (hunk 2 above)
Comment (shawbyoung) [reply bolt/include/bolt/Profile/YAMLProfileReader.h:88]: I'm moving the different matching techniques into separate functions because it'll be easier to understand and prevent the YAMLProfileReader::readProfile function from getting abhorrently large as I'll be adding call graph function matching to it in a subsequent PR. I'll add this explanation to the description.

File: bolt/lib/Profile/YAMLProfileReader.cpp [hunk 3]
```diff
@@ -334,6 +334,13 @@ Error YAMLProfileReader::preprocessProfile(BinaryContext &BC) {
   return Error::success();
 }
 
+bool YAMLProfileReader::profileMatches(
+    const yaml::bolt::BinaryFunctionProfile &Profile, BinaryFunction &BF) {
```
Comment (aaupov) [inline bolt/lib/Profile/YAMLProfileReader.cpp:338]:
```suggestion
    const yaml::bolt::BinaryFunctionProfile &Profile, const BinaryFunction &BF) {
```

**Review by aaupov — APPROVED** [review]
LG % nit

File: bolt/lib/Profile/YAMLProfileReader.cpp [hunk 4]
```diff
@@ -456,6 +435,39 @@ Error YAMLProfileReader::readProfile(BinaryContext &BC) {
...
   }
+  return MatchedWithLTOCommonName;
+}
+
+Error YAMLProfileReader::readProfile(BinaryContext &BC) {
+  if (opts::Verbosity >= 1) {
+    outs() << "BOLT-INFO: YAML profile with hash: ";
+    switch (YamlBP.Header.HashFunction) {
+    case HashFunction::StdHash:
+      outs() << "std::hash\n";
```
Comment (ayermolo) [inline bolt/lib/Profile/YAMLProfileReader.cpp:446]: BC.outs()

File: bolt/lib/Profile/YAMLProfileReader.cpp [hunk 5]
```diff
@@ -456,6 +435,39 @@ Error YAMLProfileReader::readProfile(BinaryContext &BC) {
...
+    }
+  } else if (!opts::IgnoreHash) {
+    for (BinaryFunction *BF : ProfileBFs) {
+      if (!BF)
+        continue;
+      BF->computeHash(YamlBP.Header.IsDFSOrder, YamlBP.Header.HashFunction);
+    }
+  }
+
+  size_t MatchedWithExactName = matchWithExactName();
```
Comment (ayermolo) [inline bolt/lib/Profile/YAMLProfileReader.cpp:468]: This doesn't look like it's used anywhere?

Comment (shawbyoung) [reply bolt/lib/Profile/YAMLProfileReader.cpp:468]: In lines 481 - 487 
> if (opts::Verbosity >= 1) {
>    outs() << "BOLT-INFO: matched " << MatchedWithExactName
>           << " functions with identical names\n";
> ...
Match counts are directed to outs()

File: bolt/lib/Profile/YAMLProfileReader.cpp (hunk 4 above)
Comment (shawbyoung) [reply bolt/lib/Profile/YAMLProfileReader.cpp:446]: I'm erring on the side of making minimal code change - although it's showing up on gh as code added, I haven't touched the prologue of readProfile. If you see the large "deleted" section above (starting on the original line 353 of YAMLProfileReader.cpp) it's the exact same. So, I'd like to keep this PR just about refactoring function matching.

Comment (aaupov) [reply bolt/lib/Profile/YAMLProfileReader.cpp:446]: @ayermolo, we didn't switch Profile component to BC logger class. That would be a separate effort.

File: bolt/lib/Profile/YAMLProfileReader.cpp [hunk 6]
```diff
@@ -456,6 +435,39 @@ Error YAMLProfileReader::readProfile(BinaryContext &BC) {
...
+  size_t MatchedWithHash = matchWithHash(BC);
+  size_t MatchedWithLTOCommonName = matchWithLTOCommonName();
```
Comment (maksfb) [inline bolt/lib/Profile/YAMLProfileReader.cpp:470]: nit: make them `const`.

**Review by dcci — APPROVED** [review]

**Review by maksfb — APPROVED** [review]
